From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Haochen Tong <i@hexchain.org>
Date: Mon, 11 Oct 2021 11:40:56 +0800
Subject: [PATCH] Check for libatomic dependency for atomic operations

Some platforms (e.g. RISC-V) require linking against libatomic for some
(e.g. sub-word-sized) atomic operations.

Fixes #19119.

(cherry picked from commit 57263fc10503b9875b01b4ca26ded20bba1007db)
---
 configure.ac        | 26 ++++++++++++++++++++++++++
 rts/package.conf.in |  3 +++
 rts/rts.cabal.in    |  5 +++++
 3 files changed, 34 insertions(+)

diff --git a/configure.ac b/configure.ac
index 20911ae43f..69e67fa223 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1286,6 +1286,32 @@ AC_LINK_IFELSE([
   AC_MSG_RESULT(no)
 )
 
+AC_MSG_CHECKING(whether -latomic is needed for sub-word-sized atomic operations)
+AC_LINK_IFELSE([AC_LANG_PROGRAM([[unsigned char a;]], [[__atomic_fetch_or(&a, 1, __ATOMIC_RELAXED);]])],
+    [
+        AC_MSG_RESULT(no)
+        AC_SUBST([CabalNeedLibatomic],[False])
+        need_latomic=0
+    ],
+    [
+        _save_LIBS="$LIBS"
+        LIBS="-latomic"
+        AC_LINK_IFELSE([AC_LANG_PROGRAM([[unsigned char a;]], [[__atomic_fetch_or(&a, 1, __ATOMIC_RELAXED);]])],
+            [
+                AC_MSG_RESULT(yes)
+                AC_SUBST([CabalNeedLibatomic],[True])
+                need_latomic=1
+            ],
+            [
+                AC_SUBST([CabalNeedLibatomic],[False])
+                AC_MSG_ERROR([sub-word-sized atomic operations not available.])
+                need_latomic=0
+            ])
+        LIBS="$_save_LIBS"
+    ])
+AC_DEFINE_UNQUOTED([NEED_ATOMIC_LIB], [$need_latomic],
+    [Define to 1 if we need -latomic for sub-word atomic operations.])
+
 dnl ** check for eventfd which is needed by the I/O manager
 AC_CHECK_HEADERS([sys/eventfd.h])
 AC_CHECK_FUNCS([eventfd])
diff --git a/rts/package.conf.in b/rts/package.conf.in
index 5e0af88fb6..b2d9120184 100644
--- a/rts/package.conf.in
+++ b/rts/package.conf.in
@@ -63,6 +63,9 @@ extra-libraries:
                              , "elf"
                              , "dw"             /* for backtraces */
 #endif
+#if NEED_ATOMIC_LIB
+                             , "atomic"
+#endif
 
 #if defined(INSTALLING)
 include-dirs:           INCLUDE_DIR FFI_INCLUDE_DIR
diff --git a/rts/rts.cabal.in b/rts/rts.cabal.in
index 90325dc689..5bebb4718f 100644
--- a/rts/rts.cabal.in
+++ b/rts/rts.cabal.in
@@ -20,6 +20,8 @@ flag libbfd
   default: @CabalHaveLibbfd@
 flag mingwex
   default: @CabalMingwex@
+flag need-atomic
+  default: @CabalNeedLibatomic@
 flag libdw
   default: @CabalHaveLibdw@
 flag libnuma
@@ -126,6 +128,9 @@ library
     if flag(need-pthread)
        -- for pthread_getthreadid_np, pthread_create, ...
        extra-libraries: pthread
+    if flag(need-atomic)
+       -- for sub-word-sized atomic operations (#19119)
+       extra-libraries: atomic
     if flag(libbfd)
        -- for debugging
        extra-libraries: bfd iberty
-- 
2.34.1

